tweak: optimize resource

mahanmoulaei opened the pull request:
Optimized and modified the CPU usage from **0.07**_ms_ to **0.01**_ms_ while idle.

p.s. I might do another PR in the future for **0.0**_ms_.

mahanmoulaei commented:
Those file changes weren't unnecessary because that's how I managed to lower the CPU consumption.

e.g. for ragdoll keybind, if you turn that to true in config, it would add 2 threads with loops inside running every frame, leading to even more than 0.07ms CPU usage; but, in my PR as you can see although I have turned RagdollKeybind to true in config, I'm claiming that the resource runs at **0.01**_ms_.

TayMcKenzieNZ commented:
Yes, but you done other things too, like changing menu key etc. I appreciate what you're trying to do but I would perhaps suggest forking my version up to how it is now, then optimizing it so that nothing has drastically changed.

Until then, I'm unfortunately not comfortable with accepting the PR and will have second, third, possibly even fourth opinions from other people before accepting it. 

Thanks for your understanding.

mahanmoulaei commented:
I understand your concern.
Sure ask for others' opinions as well.

Can you please be more clear about this "other things" that I've changed?

If you mean why did I change Config.FavKeybind from 171 to 'CAPITAL', I have to say they are the same keybind! so no changes are made in the menu key at all!

I might be understanding you incorrectly, and if I do, tell me...

AvaN0x commented:
Hello, I added few reviews to the changes.

> If you mean why did I change Config.FavKeybind from 171 to 'CAPITAL' and Config.RagdollKeybind from 303 to 'U', I have to say they are the same keybind! so no changes are made in the menu key at all!
I explained that to Tay, he was confused but now it's clear.

TayMcKenzieNZ commented:
Yes, apologies. Was confused lol.

mahanmoulaei commented:
> Yes, apologies. Was confused lol.

No worries Tay.

mahanmoulaei commented:
> I explained that to Tay, he was confused but now it's clear.

It's ok thanks

----
> added few reviews to the changes.

I don't see them

AvaN0x commented:
> I don't see them

Can you see them now?

mahanmoulaei commented:
> > I don't see them
> 
> Can you see them now?

Yes

mahanmoulaei commented on client/Emote.lua:
```diff
@@ -39,13 +36,21 @@ Citizen.CreateThread(function()
 end)
 
+if Config.EnableXtoCancel then
+	RegisterKeyMapping("emotecancel", "Cancel current emote", "keyboard", "X")
```
I'm aware of the ```"e c" ``` and ```"emote c" ```. The decision is up to you to decide whether you want the new ```"emotecancel"``` command or not.

mahanmoulaei commented on client/Emote.lua:
```diff
@@ -18,13 +18,10 @@ local PtfxWait = 500
-        end
 
-        if PtfxPrompt then
+        if PtfxPrompt or IsInAnimation then
```
I'll look into that

mahanmoulaei commented on client/Emote.lua:
```diff
@@ -39,13 +36,21 @@ Citizen.CreateThread(function()
+                EmoteCancel()
+            end
+        else
+            Wait(1000)
```
I'll try to come up with a better way to handle this, and if I can't, I'll lower the wait value.

mahanmoulaei commented on client/Ragdoll.lua:
```diff
@@ -1,24 +1,24 @@
-    end
-end)
+if Config.RagdollEnabled then
+	RegisterCommand('+ragdoll', function(source, args, raw) Ragdoll() end)
+	
+	RegisterCommand('-ragdoll', function(source, args, raw) StopRagdoll() end)
+	
+	RegisterKeyMapping("+ragdoll", "Ragdoll your character", "keyboard", Config.RagdollKeybind)
```
Do you want me to add Config.ToggleRagdoll in config instead of removing this feature?

AvaN0x replied:
You could make a config for it to be a toggle or not if you want. It would have to be as true by default (for consistency). I would name it something more like `RagdollAsToggle`

mahanmoulaei commented:
Could you check my comments on your review changes?

Also, I have 2 questions:

1- Do you want me to add an built-in key system so it handles all of the menu key press and helps in loops optimization? (something like [this](https://github.com/JolbakLife-RolePlay/JLRP-Framework/blob/master/client/modules/keymanager.lua))

2- Do you want me to add an optional built-in inventory config that based on the inventory system that is running, detects the player's weapon on change? This would result in a better-optimized solution for  ```   "if you start an emote and you could be able to shoot for up to a second before the IsPedShooting() check is done for the first time."```

AvaN0x commented:
> Could you check my comments on your review changes?

Done ✅ 

---

> 1- Do you want me to add an built-in key system so it handles all of the menu key press and helps in loops optimization? (something like [this](https://github.com/JolbakLife-RolePlay/JLRP-Framework/blob/master/client/modules/keymanager.lua))

I'm not a huge fan of that, because it'd add a whole lot of key to the settings and would also make the menu unusable on controllers (without adding these keymappings too, which would make again more keys in the settings).

---

> 2- Do you want me to add an optional(standalone) built-in inventory config that based on the inventory system that is running, detects the player's weapon on change? This would result in a better-optimized solution for ` "if you start an emote and you could be able to shoot for up to a second before the IsPedShooting() check is done for the first time."` | If no inventory system is found or config.inventory is set to false, the loop should run.

I'm not a huge fan of that too, because it'd add a lot of code for not that much, also it'd require to, any time a new "inventory system" gets supported, add support for it so it can get better performances. 
The best way for me would be to only start the loop when the emote starts (function `OnEmotePlay`).

mahanmoulaei commented:
> I'm not a huge fan of that, because it'd add a whole lot of key to the settings and would also make the menu unusable on controllers (without adding these keymappings too, which would make again more keys in the settings). Even more keys to the settings if other resources does that too.

RegisterKeyMapping supports controller inputs as well, so the menu being unusable on controllers can be fixed.

----
> The best way for me would be to only start the loop when the emote starts (function `OnEmotePlay`)

Alright, I'll get into it. This would work too.

mahanmoulaei commented:
Untested....

TayMcKenzieNZ commented:
Would be nice to have the controllers work 👀 Currently they work for menu navigation, however selecting and playing the animation, it doesn't do anything, prop appears but you're just standing there brainless doing nothing lol

mahanmoulaei commented:
> Would be nice to have the controllers work 👀 Currently they work for menu navigation, however selecting and playing the animation, it doesn't do anything, prop appears but you're just standing there brainless doing nothing lol

I'll look into this too

mahanmoulaei commented:
Ok, I applied the review changes.
The controller works for me in the menu now.
Let me know if you need anything else.

AvaN0x commented:
I just fixed the RagdollAsToggle config:
if true: the toggle will work
else: the player will ragdoll as long as the key is pressed

Also, you can't register a keymapping with a `+` without having the command with the `-`, because it'll trigger the command anyway which will:
- add the message in the chat if possible
- if not possible, could still make the chat show without anything new

mahanmoulaei commented:
> I just fixed the RagdollAsToggle config: if true: the toggle will work else: the player will ragdoll as long as the key is pressed
> 
> Also, you can't register a keymapping with a `+` without having the command with the `-`, because it'll trigger the command anyway which will:
> 
> * add the message in the chat if possible
> * if not possible, could still make the chat show without anything new
> 
> Also I added a new review above.

Good 👍 
I don't see your new review.

AvaN0x commented:
> I don't see your new review.

https://github.com/TayMcKenzieNZ/dpemotes/pull/40#pullrequestreview-1085089797
Can you see it with a direct link ?

mahanmoulaei commented:
> > I don't see your new review.
> 
> [#40 (review)](https://github.com/TayMcKenzieNZ/dpemotes/pull/40#pullrequestreview-1085089797) Can you see it with a direct link ?

No unfortunately

AvaN0x commented:
> > > I don't see your new review.
> > 
> > [#40 (review)](https://github.com/TayMcKenzieNZ/dpemotes/pull/40#pullrequestreview-1085089797) Can you see it with a direct link ?
> 
> No unfortunately

Well that's strange :thinking:.
Anyway here is the text for the review in Emote.lua:
> I think this call to RunAnimationThread is not needed at line 496 as the one at line 510 will be called anyway ?

mahanmoulaei commented:
> Well that's strange 🤔.
 
 Yeah
 
 ----
 
 > I think this call to RunAnimationThread is not needed at line 496 as the one at line 510 will be called anyway ?

I've coded RunAnimationThread in a way that it won't affect the performance and code block, so this means you can call this function as many times as you like without breaking the code and its performance. So no worries about RunAnimationThread being called extra/multiple times.

Tho the real question is do we need to call RunAnimationThread in line 496 while line 510 will call it anyway? The simple answer is **NO**! But since we are triggering server event of ```"dpemotes:ptfx:sync"``` in line 498, for a better sync of ptfx effects between your player and other players it's better to call RunAnimationThread and activate the loop earlier than line 510.

AvaN0x commented:
> I've coded RunAnimationThread in a way that it won't affect the performance and blocks the code, so this means you can call this function as many times as you like without breaking the code and affecting its performance. So no worries about RunAnimationThread being called extra/multiple times.

Yeah I saw that, that's how it should be anyway :)

> Tho the real question is do we **_need_** to call RunAnimationThread in line 496 while line 510 will call it anyway? The simple answer is **NO**! But since we are triggering server event of `"dpemotes:ptfx:sync"` in line 498, for a better sync of ptfx effects between your player and other players it's better to call RunAnimationThread and activate the loop earlier than line 510.

`"dpemotes:ptfx:sync" will only sync the statebags and nothing more. In my opinion the player should be able to start the ptfx (in the RunAnimationThread  thread) only if the emote has been started (line 482 with TaskPlayAnim).  So I think that we do not need both calls to RunAnimationThread, but having both of them wont break anything.

---

I'll merge your PR tomorrow I think ;)

mahanmoulaei commented on client/Emote.lua:
```diff
@@ -470,6 +493,7 @@ function OnEmotePlay(EmoteName)
             PtfxCanHold = EmoteName.AnimationOptions.PtfxCanHold
             PtfxNotif = false
             PtfxPrompt = true
+            RunAnimationThread()
```
I see the review changes now lol

mahanmoulaei commented:
> the player should be able to start the ptfx (in the RunAnimationThread  thread) only if the emote has been started

This also makes sense too. I guess you'll have to test both of them to see which way gives you the most desired output in-game, and then apply that in code.

--------

> I'll merge your PR tomorrow I think ;)

Perfecttt 😉

TayMcKenzieNZ commented:
Thanks guys 😊

AvaN0x commented:
Merged :)

MichaelCoding25 commented on client/Ragdoll.lua:
```diff
@@ -1,24 +1,28 @@
-        if isInRagdoll then
-            SetPedToRagdoll(PlayerPedId(), 1000, 1000, 0, 0, 0, 0)
+    RegisterCommand('+ragdoll', function(source, args, raw) Ragdoll() end)
+    egisterCommand('-ragdoll', function(source, args, raw) StopRagdoll() end)
```
Well this won't work lol

AvaN0x replied:
Oh, thank you for point it out xD
fixed in 5185d46

TayMcKenzieNZ commented:
Controller works 🥳 Which is weird considering nothing was really changed in that regards lol

Thanks for assisting us 🙂
